### PR TITLE
Updates after july release

### PR DESCRIPTION
**Pull request by ibnt1**
- resolve destinations references 
- split line logic rewritten to not use snap+split, which causes incorrect split artifacts in some edge cases. instead assign to each new split point the corresponding coordinate index from the original geometry and use those to reconstruct geometries for each split
- apply lr logic revisited
- added basic metrics to help analyze potential data issues after split:
   - "bucket" of segment length - very small (centimeter length) could be a sign of LR value problems that could be addressed with tweaking threshold
   - self intersecting roads - before and after split
- tests 
   - moved to separate files
   - added more edge cases for splits, consolidated code; 
   - added tests for apply_lr
- added GH action to run tests
- add requirements.txt, vscode launcher cfg
- updates to doc, paths

### CODE REVIEW

File: TransportationSplitter.py [hunk 1]
```diff
@@ -190,31 +195,26 @@ def are_different_coords(coords1, coords2):
...
-        snapped_geometry = ops.snap(remaining_line_geometry, next_split_geometry, tolerance=0.0000001)
-        split_line = ops.split(snapped_geometry, next_split_geometry)
-        split_segment_geometry = remove_consecutive_dupe_coords(LineString(split_line.geoms[0]))
-        if len(split_line.geoms) > 1:
-            # Only use remainder of line for snapping to minimize chance of snapping to the wrong location when segment is self-intersecting
-            remaining_line_geometry = split_line.geoms[1]
-        split_segments.append(SplitSegment(i, split_segment_geometry, split_points[i], split_points[i + 1]))
-
+    i=0
+    for split_point_start, split_point_end in zip(split_points[:-1], split_points[1:]):
```
Comment (brad-richardson) [inline TransportationSplitter.py:218]: Maybe just use a index/range iterator for this? You're already tracking an index with `i`

File: TransportationSplitter.py [hunk 2]
```diff
@@ -190,31 +195,26 @@ def are_different_coords(coords1, coords2):
...
+        idx_start = split_point_start.at_coord_idx + 1
+        idx_end = split_point_end.at_coord_idx + 1
```
Comment (brad-richardson) [inline TransportationSplitter.py:220]: Why the +1 on these? Shouldn't the second one be -1 as well?

File: README.md [hunk 3]
```diff
@@ -28,12 +28,13 @@ Any Spark environment should work, but for reference this was tested on Databric
 
 ### Dependencies
 
-pip install apache-sedona shapely
+pip install apache-sedona shapely pyproj
```
[the remark this answers is not in the document]
Comment (ibnt1) [reply README.md:31]: yes, it is imported https://github.com/OvertureMaps/transportation-splitter/blob/2d801266ed88f3ae1a21a018395317add9d261b0/TransportationSplitter.py#L14

File: TransportationSplitter.py [hunk 4]
```diff
@@ -376,56 +497,130 @@ def add_lr_split_points(split_points, lrs, segment_id, original_segment_geometry
     # Get the length of the projected segment
     geod = pyproj.Geod(ellps="WGS84")
     line_length = geod.geometry_length(original_segment_geometry)
```
Comment (brad-richardson) [inline TransportationSplitter.py:499]: Switch this to use `get_length` as well?

**Review by brad-richardson — APPROVED** [review]
🚀🚀🚀

File: TransportationSplitter.py [hunk 5]
```diff
@@ -349,12 +462,20 @@ def get_split_segment_dict(original_segment_dict, split_segment, lr_columns_for_
         prohibited_transitions = modified_segment_dict.get(PROHIBITED_TRANSITIONS_COLUMN) or {}
         modified_segment_dict[PROHIBITED_TRANSITIONS_COLUMN], modified_segment_dict["turn_restrictions"] = get_trs(prohibited_transitions, modified_segment_dict["connector_ids"])
 
+    if DESTINATIONS_COLUMN in original_segment_dict:
```
Comment (brad-richardson) [inline TransportationSplitter.py:465]: I think you also need `DESTINATIONS_COLUMN in lr_columns_for_splitting` here

File: TransportationSplitter.py [hunk 6]
```diff
@@ -506,38 +721,41 @@ def split_segment(input_segment):
...
+            add_lr_split_points(split_points, lrs, original_segment_dict["id"], input_segment.geometry, segment_length)
 
             sorted_split_points = sorted(split_points, key=lambda p: p.lr)
             debug_messages.append("sorted final split points:")
             for p in sorted_split_points:
-                debug_messages.append(", ".join([p.id, str(p.geometry), str(p.lr)]))
+                debug_messages.append(str(p))
 
             if len(sorted_split_points) < 2:
                 raise Exception(f"Unexpected number of split points: {str(len(sorted_split_points))}; (expected at least 2)")
```
Comment (brad-richardson) [inline TransportationSplitter.py:750]: It looks like I'm getting a bunch of failures here when trying to not split on connectors

Comment (brad-richardson) [reply TransportationSplitter.py:750]: I think above instead of having split_points be empty, we should do this instead:
```
            if not SPLIT_AT_CONNECTORS:
                # Only get first and last, downstream logic is dependent on that, ignore other connectors
                split_points = [split_points[0], split_points[-1]]
```
I'm testing that now

File: TransportationSplitter.py [hunk 7]
```diff
@@ -467,12 +662,32 @@ def make_split_point(connector, segment_geometry):
...
+        StructField('final_heading', StringType(), True)
+        ]), False)
+
 def split_joined_segments(df: DataFrame, lr_columns_for_splitting: list[str]) -> DataFrame:
     broadcast_lr_columns_for_splitting = sc.broadcast(lr_columns_for_splitting)    
     input_fields_to_drop_in_splits = ["joined_connectors"]
     transportation_feature_schema = StructType([field for field in df.schema.fields if field.name not in input_fields_to_drop_in_splits])
     split_segment_fields = [field for field in transportation_feature_schema] + additional_fields_in_split_segments
-    if PROHIBITED_TRANSITIONS_COLUMN in lr_columns_for_splitting:
+    if PROHIBITED_TRANSITIONS_COLUMN in df.columns:
```
[the remark this answers is not in the document]
Comment (ibnt1) [reply TransportationSplitter.py:690]: this _if in  lr_columns_for_splitting_ was wrong, the prohibited_transitions references need to be resolved to have start/end lr no matter what, because the target segment ids do get split.

File: TransportationSplitter.py (hunk 4 above)
Comment (ibnt1) [reply TransportationSplitter.py]: good catch! length is actually passed on as it's already calculated by the caller, i missed this

File: TransportationSplitter.py (hunk 6 above)
Comment (ibnt1) [reply TransportationSplitter.py:749]: yeah, that's probably simplest workaround, i'm running a test with that too.

but i'm wondering if this should be though out a bit more, for example it is arguably preferable to reuse existing connectors as much as possible if they exists to correspond to property LR values, rather than introducing new "artificial" connectors like we would with this approach for split_at_connectors=false.
so alternative would be to keep splitpoint objects for all connecters, try to snap LR values to them, create new splitpoints for LRs only when they are not within the "snap" threshold, but then only split at the split points that have lrs  values snapped to them for which the user wants to introduce splits. logic is a bit hairy but would reduce the amount of new connectors that it is adding.

Comment (ibnt1) [reply TransportationSplitter.py:751]: included the workaround for now, test worked fine

File: README.md [hunk 8]
```diff
@@ -28,12 +28,13 @@ Any Spark environment should work, but for reference this was tested on Databric
...
 
 Following packages versions were used:
 ```
 apache-sedona = "1.5.1"
 shapely = "2.0.2"
+pyproj = "3.6.1"
```
Comment (bastiaanvanassche-tomtom) [inline README.md:37]: maybe omit the specific version of the package over here, as these versions can quickly become out of sync with actual version being used in the `requirements.txt`:  the `apache-sedona` version defined in the `requirements.txt` is already different from the one here in the `README.md`

File: .github/workflows/tests.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    branches: [main]
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.9, 3.11]
+    steps:
+    - name: Check out repository
+      uses: actions/checkout@v3
```
Comment (bastiaanvanassche-tomtom) [inline .github/workflows/tests.yaml:15]: there's already a `4.x` series GH action available for `actions/checkout`: https://github.com/actions/checkout/releases/tag/v4.1.7

File: .github/workflows/tests.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v4
```
Comment (bastiaanvanassche-tomtom) [inline .github/workflows/tests.yaml:17]: there's already a `5.x` series GH action available for `actions/setup-python`: https://github.com/actions/setup-python/releases/tag/v5.1.1

File: .vscode/settings.json [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    "python.testing.unittestArgs": [
+        "-v",
+        "-s",
+        ".",
+        "-p",
+        "test_*.py"
+    ],
+    "python.testing.pytestEnabled": false,
+    "python.testing.unittestEnabled": true
+}
```
Comment (bastiaanvanassche-tomtom) [inline .vscode/settings.json:11]: missing EOL character

File: .vscode/launch.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    "configurations": [
+        {
+            "name": "Python Debugger: Current File",
+            "type": "debugpy",
+            "request": "launch",
+            "program": "${file}",
+            "console": "integratedTerminal"
+        }
+    ]
+}
```
Comment (bastiaanvanassche-tomtom) [inline .vscode/launch.json:12]: missing EOL character

File: .gitignore [hunk 13]
```diff
@@ -102,3 +102,7 @@ dist
 
 # TernJS port file
 .tern-port
+
+.venv
+
+__pycache__
```
Comment (bastiaanvanassche-tomtom) [inline .gitignore:108]: missing EOL character

File: test_apply_lr.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
+import unittest
+from TransportationSplitter import *
+
+class TestApplyLR(unittest.TestCase):
```
Comment (bastiaanvanassche-tomtom) [inline test_apply_lr.py:4]: Not sure if we want to use `pytest` instead of `unittest`?
I believe pytest is more feature-rich with the drawback that it is not included in Python Standard Library.
Didn't we also make the switch to `pytest` in the combobulator lately?

File: .github/workflows/tests.yaml [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Check Python version
+      run: python --version
+    - name: Upgrade pip
+      run: pip install --upgrade pip
+    - name: Install packages
+      run: pip install -r requirements.txt
+    - name: Run tests
+      run: python -m unittest discover -s . -p "test_*.py" -v
```
Comment (bastiaanvanassche-tomtom) [inline .github/workflows/tests.yaml:27]: missing EOL character

File: TransportationSplitter.py
Comment (bastiaanvanassche-tomtom) [inline TransportationSplitter.py:1]: not sure if we want to adhere to PEP8 regarding module naming, https://peps.python.org/pep-0008/#package-and-module-names. I would assume the module/file needs to be named `transportation-splitter.py`

File: TransportationSplitter.py [hunk 16]
```diff
@@ -329,18 +395,65 @@ def get_trs(turn_restrictions, connector_ids):
...
+        return False
+    if when_heading == "backward" and connector_ids[0] != from_connector_id:
+        # the first connector id on this segment split needs to match the from_connector_id in the destination because heading scope applies only to backward
+        return False
+    if from_connector_id not in connector_ids:
+        # the from_connector_id needs to be in this split's connector_ids no matter what
+        return False
+    return True
+
+def get_length_bucket(length):
```
Comment (bastiaanvanassche-tomtom) [inline TransportationSplitter.py:424]: not sure which versions of python we support but this method could be an ideal candidate for a Structural Pattern Matching expression, https://stackoverflow.com/a/11479840 and https://peps.python.org/pep-0634/
it's only supported as of python `3.10`...

File: test_apply_lr.py (hunk 14 above)
Comment (ibnt1) [reply test_apply_lr.py:4]: that's true, pytest is richer in features; brad and i discussed to stick with unittest for now as we're not really needing those features and we want to limit number of package dependencies

File: TransportationSplitter.py (hunk 16 above)
Comment (ibnt1) [reply TransportationSplitter.py:424]: as discussed, we'll probably want to avoid things that limit the versions this can run on

File: .github/workflows/tests.yaml (hunk 9 above)
Comment (ibnt1) [reply .github/workflows/tests.yaml]: bumped!

File: .github/workflows/tests.yaml (hunk 10 above)
Comment (ibnt1) [reply .github/workflows/tests.yaml]: bumped!

**Comment by ibnt1** [comment]
@bastiaanvanassche-tomtom i'll ignore the whitespace stuff for now, since PR https://github.com/OvertureMaps/transportation-splitter/pull/7 is fixing and enforcing it

File: TransportationSplitter.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ibnt1) [reply TransportationSplitter.py:689]: this should be fixed now

File: TransportationSplitter.py
Comment (ibnt1) [reply TransportationSplitter.py:1]: done, renamed

File: README.md (hunk 8 above)
Comment (ibnt1) [reply README.md]: makes sense, i replaced it with link to requirements instead

**Review by bastiaanvanassche-tomtom — APPROVED** [review]
